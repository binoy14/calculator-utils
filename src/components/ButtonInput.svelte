<script lang="ts">
  import { createEventDispatcher } from 'svelte';
  import CloseIcon from './CloseIcon.svelte';

  const dispatch = createEventDispatcher();

  export let items: number[];
  export let label: string;
  export let value: number | undefined;
  export let title: string;

  let handleCalculation = (val?: number) => {
    dispatch('inputChange', val);
  };

  $: {
    handleCalculation(value);
  }

  function handleClear() {
    handleCalculation(undefined);
  }
</script>

<div class="mt-4 w-full shadow">
  <div class="w-full rounded-t border bg-white pl-2 text-amber-700">{title}:</div>
  <div class="inline-flex w-full">
    {#each items as item, i}
      <button
        class="bg-white px-2 py-4 font-bold shadow"
        class:first-item={i === 0}
        class:active={value === item}
        on:click={() => handleCalculation(item)}>{item}%</button
      >
    {/each}
    <div class="inline-flex rounded-br shadow">
      <label for={label} class="sr-only">{title}</label>
      <input type="number" id={label} bind:value placeholder="%" class="w-full appearance-none px-2 py-4" min="0" />
      <button class="bg-white px-2" on:click={handleClear}>
        <CloseIcon />
      </button>
    </div>
  </div>
</div>

<style lang="postcss">
  input[type='number']::-webkit-inner-spin-button,
  input[type='number']::-webkit-outer-spin-button {
    @apply appearance-none;
  }

  .first-item {
    @apply rounded-bl;
  }

  .active {
    @apply bg-gray-200;
  }
</style>
