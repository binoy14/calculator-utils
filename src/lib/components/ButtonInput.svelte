<script lang="ts">
  import { createEventDispatcher } from 'svelte';
  import CloseIcon from './CloseIcon.svelte';

  const dispatch = createEventDispatcher();

  export let items: number[];
  export let label: string;
  export let value: number | undefined;
  export let title: string;
  export let placeholder: string = '';
  export let formatter: (value: number) => string;

  let handleCalculation = (val?: number) => {
    dispatch('inputChange', val);
  };

  $: {
    handleCalculation(value);
  }

  function handleClear() {
    handleCalculation(undefined);
  }
</script>

<div class="mt-4 w-full shadow">
  <div class="w-full rounded-t border bg-white pl-2 text-amber-700">{title}:</div>
  <div class="inline-flex w-full">
    {#each items as item, i}
      <button
        class="border-r bg-white px-2 py-4 font-bold"
        class:first-item={i === 0}
        class:active={value === item}
        data-testid={`${label}-${item}`}
        on:click={() => handleCalculation(item)}>{formatter(item)}</button
      >
    {/each}
    <div class="inline-flex">
      <label for={label} class="sr-only">{title}</label>
      <input
        type="number"
        id={label}
        bind:value
        {placeholder}
        class="w-full appearance-none px-2 py-4"
        min="0"
      />
      <button class="rounded-br bg-white px-2" on:click={handleClear}>
        <CloseIcon />
      </button>
    </div>
  </div>
</div>

<style lang="postcss">
  .first-item {
    @apply rounded-bl;
  }

  .active {
    @apply bg-gray-200;
  }
</style>
