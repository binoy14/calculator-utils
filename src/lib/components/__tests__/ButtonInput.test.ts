import { render } from '@testing-library/svelte';

import ButtonInput from '../ButtonInput.svelte';

test('renders the button input', () => {
  const component = render(ButtonInput, {
    props: {
      items: [1],
      label: 'button',
      title: 'Button',
      value: 1,
      placeholder: '%',
      formatter: (value: number) => `${value}`,
    },
  });

  expect(component.container).toMatchSnapshot();
});
