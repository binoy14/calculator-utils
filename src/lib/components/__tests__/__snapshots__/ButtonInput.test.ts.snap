// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the button input 1`] = `
<div>
  <div
    class="mt-4 w-full shadow"
  >
    <div
      class="w-full rounded-t border bg-white pl-2 text-amber-700"
    >
      Button
      :
    </div>
     
    <div
      class="inline-flex w-full"
    >
      <button
        class="border-r bg-white px-2 py-4 font-bold svelte-1kvtb7z first-item active"
        data-testid="button-1"
      >
        1
      </button>
       
      <div
        class="inline-flex"
      >
        <label
          class="sr-only"
          for="button"
        >
          Button
        </label>
         
        <input
          class="w-full appearance-none px-2 py-4"
          id="button"
          min="0"
          placeholder="%"
          type="number"
        />
         
        <button
          class="rounded-br bg-white px-2"
        >
          <svg
            class="svelte-r5w73o"
            height="18"
            viewBox="0 -960 960 960"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m330-288 150-150 150 150 42-42-150-150 150-150-42-42-150 150-150-150-42 42 150 150-150 150 42 42ZM480-80q-82 0-155-31.5t-127.5-86Q143-252 111.5-325T80-480q0-83 31.5-156t86-127Q252-817 325-848.5T480-880q83 0 156 31.5T763-763q54 54 85.5 127T880-480q0 82-31.5 155T763-197.5q-54 54.5-127 86T480-80Zm0-60q142 0 241-99.5T820-480q0-142-99-241t-241-99q-141 0-240.5 99T140-480q0 141 99.5 240.5T480-140Zm0-340Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
